core/src/main/scala/filodb.core/downsample/DownsampledTimeSeriesStore.scala

```diff
@@ -0,0 +1,175 @@
+                       chunkMethod: ChunkScanMethod): PartLookupResult = {
+    val shard = datasets(ref).get(partMethod.shard)
+
+    if (shard == UnsafeUtils.ZeroPointer) {
+      throw new IllegalArgumentException(s"Shard $shard of dataset $ref is not assigned to " +
+        s"this node. Was it was recently reassigned to another node? Prolonged occurrence indicates an issue.")
+    }
+    shard.lookupPartitions(partMethod, chunkMethod)
+  }
+
+  def scanPartitions(ref: DatasetRef,
+                     lookupRes: PartLookupResult): Observable[ReadablePartition] = {
+    val shard = datasets(ref).get(lookupRes.shard)
+
+    if (shard == UnsafeUtils.ZeroPointer) {
+      throw new IllegalArgumentException(s"Shard $shard of dataset $ref is not assigned to " +
+        s"this node. Was it was recently reassigned to another node? Prolonged occurrence indicates an issue.")
+    }
+    shard.scanPartitions(lookupRes)
+
+  }
+
+  def shardMetrics(dataset: DatasetRef, shard: Int): TimeSeriesShardStats =
+    getShard(dataset, shard).get.shardStats
+
+  def activeShards(dataset: DatasetRef): Seq[Int] =
+    datasets.get(dataset).map(_.keySet.asScala.map(_.toInt).toSeq).getOrElse(Nil)
+
+  def getScanSplits(dataset: DatasetRef, splitsPerNode: Int = 1): Seq[ScanSplit] =
+    activeShards(dataset).map(ShardSplit)
+
+  def groupsInDataset(ref: DatasetRef): Int = throw new UnsupportedOperationException()
+
+  def analyzeAndLogCorruptPtr(ref: DatasetRef, cve: CorruptVectorException): Unit =
+    throw new UnsupportedOperationException()
+
+  def reset(): Unit = {
+    datasets.clear()
+    store.reset()
+  }
+
+  def removeShard(dataset: DatasetRef, shardNum: Int, shard: DownsampledTimeSeriesShard): Boolean = {
+    datasets.get(dataset).map(_.remove(shardNum, shard)).getOrElse(false)
+  }
+
+  def shutdown(): Unit = {
+    reset()
+  }
+
+  override def ingest(dataset: DatasetRef, shard: Int,
+                      data: SomeData): Unit = throw new UnsupportedOperationException()
+
+  override def ingestStream(dataset: DatasetRef,
+                   shard: Int,
+                   stream: Observable[SomeData],
+                   flushSched: Scheduler,
+                   cancelTask: Task[Unit] = Task {}): CancelableFuture[Unit] = throw new UnsupportedOperationException()
+
+  override def recoverStream(dataset: DatasetRef, shard: Int,
+                             stream: Observable[SomeData],
+                             startOffset: Long, endOffset: Long, checkpoints: Map[Int, Long],
+                             reportingInterval: Long): Observable[Long] = throw new UnsupportedOperationException()
+
+  override def numPartitions(dataset: DatasetRef, shard: Int): Int = throw new UnsupportedOperationException()
+
+  override def numRowsIngested(dataset: DatasetRef, shard: Int): Long = throw new UnsupportedOperationException()
+
+  override def latestOffset(dataset: DatasetRef, shard: Int): Long = throw new UnsupportedOperationException()
```

Comment:
Isn't ??? the same as throwing this exception?

Reply:
Both throw exceptions, but different ones. `???` is technically not-implemented-error.